### PR TITLE
HARP-732: Automatic tests for berlin themes against json schema.

### PR DESCRIPTION
**Pull request by zbigg**
* fix `SelectorValueDefinition.value` typing to contain expressions
* add test for Berlin theme validity against generated JSON schema
* fix berlin themes to comply with schema

### CODE REVIEW

File: @here/harp-map-theme/test/DefaultThemeTest.ts [hunk 1]
```diff
@@ -9,8 +9,44 @@
...
-    it("ok", function() {
-        assert.isTrue(true);
+import * as Ajv from "ajv";
+import * as fs from "fs";
+import * as path from "path";
+
+const isNode = typeof window === "undefined";
+
+describe("Berlin Theme", function() {
+    if (!isNode) {
```
[the remark this answers is not in the document]
Comment (zbigg) [reply @here/harp-map-theme/test/DefaultThemeTest.ts]: Actually, i was almost sure that `ajv` is node only. Will fix it.

File: @here/harp-datasource-protocol/lib/Theme.ts [hunk 2]
```diff
@@ -228,7 +228,7 @@ export interface SelectorValueDefinition extends BaseValueDefinition {
      *
      * See [[BaseStyle.when]].
      */
-    value: string;
+    value: string | any[];
```
[the remark this answers is not in the document]
Comment (zbigg) [reply @here/harp-datasource-protocol/lib/Theme.ts]: This is not `Expr`, this JSON representation of `Expr` which (unfortunately) cannot be properly typed until /https://github.com/microsoft/TypeScript/issues/6230is solved.

Will change to `unknown[]` anyway.

[the remark this answers is not in the document]
Comment (okanishchev) [reply @here/harp-datasource-protocol/lib/Theme.ts]: Seems to be there are solution to recursive types via intermediate inherited sub-types:
https://github.com/microsoft/TypeScript/issues/6230#issuecomment-259766250

```typescript
    interface BaseConditionSpec {
        /** The Condition must match. The convention is the provide a string or array of strings here, but it's not enforced. */
        include?: Condition;
        /** The Condition must NOT match. The convention is the provide a string or array of strings here, but it's not enforced. */
        exclude?: Condition;
    }

    interface TestConditionSpec extends BaseConditionSpec {
        /** The Condition must match. The convention is the provide a RegExp or array of RegExps here, but it's not enforced. */
        test: Condition;
    }

    interface AndConditionSpec extends BaseConditionSpec {
        /** All Conditions must match. */
        and: ConditionArray;
    }

    interface OrConditionSpec extends BaseConditionSpec {
        /** Any Condition must match. */
        or: ConditionArray;
    }

    interface NotConditionSpec extends BaseCondtionSpec {
        /** The Condition must NOT match. */
        not: Condition;
    }

    interface ConditionArray extends Array<Condition> {} 
    type Condition = string | RegExp | ((absPath: string) => boolean) | TestConditionSpec | OrConditionSpec | AndConditionSpec | NotConditionSpec| ConditionArray;
```
Can we do this, or it not worth effort?

[the remark this answers is not in the document]
Comment (zbigg) [reply @here/harp-datasource-protocol/lib/Theme.ts]: @okanishchev, please look how Expr as json works (https://github.com/heremaps/harp.gl/blob/master/%40here/harp-datasource-protocol/lib/Expr.ts). It's based on arrays which emulate s-expressions, it cannot be typed now in typescript.
Even, if there are some hacks, they are ugly and/or not usable with json schema generators.

You can do it when there is interface on the way. With bare arrays (and that's how our expressions are), you can't. Please show me example, where this works:
```
type Add = ['add', ...Expression[]]
type And = ['&&', Expression, Expression]
type Literal = string | number| boolean
type Expression = Add | And | Literal;
```

[the remark this answers is not in the document]
Comment (zbigg) [reply @here/harp-datasource-protocol/lib/Theme.ts]: Tried this too:
```
type Add = ['add', ...Subexpressions] // A rest element type must be an array type.t
type And = ['&&', ...Subexpressions]
type Literal = string | number| boolean;
type Expression = Add | And | Literal; // Type alias 'Expression' circularly references itself.ts(2456)

interface Subexpressions extends Array<Expression> { }
```

(From: https://github.com/microsoft/TypeScript/issues/3496#issuecomment-128553540)

[the remark this answers is not in the document]
Comment (zbigg) [reply @here/harp-datasource-protocol/lib/Theme.ts]: And this:
```
export type Expression = Add | Any | string | number;

export interface Add extends Array<any> {
  0: 'add';
  1: Expression;
  2: Expression;
}

export interface Any extends Array<any> {
  0: 'any';
  1: Expression;
  2: Expression;
}

const y: Expression = ['add', 2, ['any', 1, 2, {}]];
```
(It looks like it works, but is _ugly_, unreadable and actually very permissive so not sure where is the benefit)

From: https://github.com/microsoft/TypeScript/issues/6230#issuecomment-259920795

[the remark this answers is not in the document]
Comment (okanishchev) [reply @here/harp-datasource-protocol/lib/Theme.ts]: I see now, you can do this with objects, but not with arrays. Typescript.

**Pull request merged**